webapp-django/crashstats/cron/management/commands/cron.py

```diff
@@ -0,0 +1,464 @@
+            return False
+
+        else:
+            return True
+
+    return job.next_run < now
+
+
+def get_run_times(job_spec, last_success):
+    """Return generator of run_times for this job."""
+    now = timezone.now()
+
+    # If this is not a backfill job, just run it
+    if not job_spec.get('backfill', False):
+        yield now
+        return
+
+    # If this is a backfill command that's never been run before, run it
+    if last_success is None:
+        yield now
+        return
+
+    # Figure out the backfill times and yield those; base it on the
+    # first_run datetime so the job doesn't drift
+    when = last_success
+    if job_spec.get('time'):
+        # So, reset the hour/minute part to always match the
+        # intention.
+        h, m = [int(x) for x in job_spec['time'].split(':')]
+        when = when.replace(
+            hour=h,
+            minute=m,
+            second=0,
+            microsecond=0
+        )
+    seconds = convert_frequency(job_spec.get('frequency'))
+    interval = datetime.timedelta(seconds=seconds)
+    # Loop over each missed interval from the time of the last success,
+    # forward by each interval until it reaches the time 'now'.
+    while (when + interval) <= now:
+        when += interval
+        yield when
+
+
+class Command(BaseCommand):
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--list-jobs', action='store_true',
+            help='List all jobs.'
+        )
+        parser.add_argument(
+            '--mark-success', type=str, default='',
+            help='Comma-delimited list of job names to mark successful or "all" for all.'
+        )
+        parser.add_argument(
+            '--reset-job', type=str, default='',
+            help='Comma-delimited list of job names to reset or "all" for all.'
+        )
+        parser.add_argument(
+            '--job', type=str, default='',
+            help='Job name for job you want to run.'
+        )
+        parser.add_argument(
+            '--force', action='store_true',
+            help='Force a job to run even if it is not time.'
+        )
+
+        # This allows us to pass in sub command arguments through the cron
+        # command
+        parser.add_argument(
+            '--job-arg', action='append',
+            help='Cron job command arguments in name=value format.'
+        )
+
+    def handle(self, **options):
+        """Execute crontabber command."""
+        if options['list_jobs']:
+            return self.cmd_list_jobs()
+        elif options['mark_success']:
+            return self.cmd_mark_success(options['mark_success'])
+        elif options['reset_job']:
+            return self.cmd_reset_job(options['reset_job'])
+        elif options['job']:
+            job_args = {}
+            for arg in (options['job_arg'] or []):
+                if '=' in arg:
+                    key, val = arg.split('=', 1)
+                else:
+                    key, val = arg, True
```

Comment:
If you do `--job-arg=key=True` and `--job-arg=key`, you get two different values of `True`. I'll have to think about that, but it seems kind of meh.

The other unenthusing thing here is that `key` is the variable name--not the arg name. So if you have an arg like:

```
parser.add_argument('--foo-bar')
```

then you would run that command with `--foo-bar`, but you'd pass in `foo_bar` here as the key.

I'll have to think about that, too.